dll/shellext/mydocs/CMyDocsDropHandler.cpp

```diff
@@ -0,0 +1,214 @@
+/*
+ * PROJECT:     mydocs
+ * LICENSE:     LGPL-2.1+ (https://spdx.org/licenses/LGPL-2.1+)
+ * PURPOSE:     MyDocs implementation
+ * COPYRIGHT:   Copyright 2020 Katayama Hirofumi MZ (katayama.hirofumi.mz@gmail.com)
+ */
+
+#include "precomp.hpp"
+
+WINE_DEFAULT_DEBUG_CHANNEL(mydocs);
+
+static CLIPFORMAT g_cfHIDA = 0;
+
+CMyDocsDropHandler::CMyDocsDropHandler()
+{
+    InterlockedIncrement(&g_ModuleRefCnt);
+}
+
+CMyDocsDropHandler::~CMyDocsDropHandler()
+{
+    InterlockedDecrement(&g_ModuleRefCnt);
+}
+
+// IDropTarget
+STDMETHODIMP
+CMyDocsDropHandler::DragEnter(IDataObject *pDataObject, DWORD dwKeyState,
+                                POINTL pt, DWORD *pdwEffect)
+{
+    TRACE("(%p)\n", this);
+
+    *pdwEffect &= DROPEFFECT_COPY; // Copy only
+
+    return S_OK;
+}
+
+STDMETHODIMP
+CMyDocsDropHandler::DragOver(DWORD dwKeyState, POINTL pt, DWORD *pdwEffect)
+{
+    TRACE("(%p)\n", this);
+
+    *pdwEffect &= DROPEFFECT_COPY; // Copy only
+
+    return S_OK;
+}
+
+STDMETHODIMP CMyDocsDropHandler::DragLeave()
+{
+    TRACE("(%p)\n", this);
+    return S_OK;
+}
+
+STDMETHODIMP
+CMyDocsDropHandler::Drop(IDataObject *pDataObject, DWORD dwKeyState,
+                         POINTL pt, DWORD *pdwEffect)
+{
+    TRACE("(%p)\n", this);
+
+    if (!pDataObject)
+    {
+        ERR("pDataObject is NULL\n");
+        *pdwEffect = 0;
+        DragLeave();
+        return E_POINTER;
+    }
+
+    CComPtr<IShellFolder> pDesktop;
+    HRESULT hr = SHGetDesktopFolder(&pDesktop);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    // get the clipboard format
+    if (g_cfHIDA == 0)
+        g_cfHIDA = ::RegisterClipboardFormatW(CFSTR_SHELLIDLIST);
+
+    // Retrieve an HIDA (handle of IDA)
+    STGMEDIUM medium;
+    FORMATETC fmt = { g_cfHIDA, NULL, DVASPECT_CONTENT, -1, TYMED_HGLOBAL };
+    hr = pDataObject->GetData(&fmt, &medium);
+    if (FAILED_UNEXPECTEDLY(hr))
+    {
+        *pdwEffect = 0;
+        DragLeave();
+        return E_FAIL;
+    }
+
+    // lock HIDA
+    LPIDA pida = reinterpret_cast<LPIDA>(GlobalLock(medium.hGlobal));
+    UINT iItem, cItems = pida->cidl;
+
+    // get the path of "My Documents"
+    WCHAR szzDir[MAX_PATH + 1];
+    SHGetSpecialFolderPathW(NULL, szzDir, CSIDL_PERSONAL, FALSE);
+    szzDir[lstrlenW(szzDir) + 1] = 0; // ends with double NULs
+
+    // for all source items
+    CStringW strSrcList;
+    WCHAR szSrc[MAX_PATH];
+    const BYTE *pb = reinterpret_cast<BYTE *>(pida);
+    PCIDLIST_ABSOLUTE pidlParent = reinterpret_cast<PCIDLIST_ABSOLUTE>(pb + pida->aoffset[0]);
+    for (iItem = 0; iItem < cItems; ++iItem)
+    {
+        // query source pidl
+        PCITEMID_CHILD pidlChild = reinterpret_cast<PCITEMID_CHILD>(pb + pida->aoffset[iItem + 1]);
```

Comment:
`HIDA_GetPIDLItem`

Reply:
Fixed in https://github.com/reactos/reactos/commit/2dc2ec9de359e4d6c31b6f61ba305e4323db8ddd.